src/common/oclengine.cpp

```diff
@@ -94,18 +103,27 @@ void OCLEngine::InitOCL()
     for (int i = 0; i < deviceCount; i++) {
         // a context is like a "runtime link" to the device and platform;
         // i.e. communication is possible
-        all_device_contexts.push_back(std::make_shared<OCLDeviceContext>(devPlatVec[i], all_devices[i]));
-        all_device_contexts[i]->context = cl::Context(all_devices[i]);
-        all_device_contexts[i]->queue = cl::CommandQueue(all_device_contexts[i]->context, all_devices[i]);
+        std::shared_ptr<OCLDeviceContext> devCntxt = std::make_shared<OCLDeviceContext>(devPlatVec[i], all_devices[i]);
+        devCntxt->context = cl::Context(all_devices[i]);
+        devCntxt->queue = cl::CommandQueue(devCntxt->context, all_devices[i]);
 
-        cl::Program program = cl::Program(all_device_contexts[i]->context, sources);
+        cl::Program program = cl::Program(devCntxt->context, sources);
 
         if (program.build({ all_devices[i] }) != CL_SUCCESS) {
             std::cout << "Error building for device #" << i << ": "
                       << program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(all_devices[i]) << std::endl;
-            exit(1);
+
+            if (i == dev) {
+                default_device_context = all_device_contexts[0];
```

Comment:
What ifthis was the first device that failed?

Reply:
The default device is set above as the last device in the list. If it's also the first device, there are no OpenCL devices, and this can't continue either way.

Reply:
I'm adding a comment to this effect.